Improve doc for unsigned transaction

gui1117 opened the pull request.

Zebedeusz commented on substrate/frame/system/src/offchain.rs:
```diff
@@ -489,8 +489,14 @@ pub trait CreateSignedTransaction<LocalCall>:
 /// Interface for creating an inherent.
+///
+/// It can also be used to create an unsigned transaction as they are both the same extrinsic
+/// variant: `Bare`. Unsigned transaction are deprecated in favor of general transaction.
```
What do you mean by that 2nd phrase?

gui1117 replied:
The extrinsic format is has 3 variants: signed, general and bare.
signed and general both have call + additional data for transaction extensions +signature (if signed, not if general), bare only has a call.

there is different kind of extrinsic: inherent, unsigned transaction: they use the bare format. signed and general transaction they use their respective format.
We want to deprecate unsigned transaction, people can already use general transaction instead, but the PR with `pallet::authorize` will help the transition.
Looking back we shouldn't have pushed the deprecation so fast.

there is some doc here: https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/extrinsic_encoding/index.html

bkchr commented on substrate/frame/system/src/offchain.rs:
```diff
@@ -489,8 +489,14 @@ pub trait CreateSignedTransaction<LocalCall>:
+///
+/// It can also be used to create an unsigned transaction as they are both the same extrinsic
+/// variant: `Bare`. Unsigned transaction are deprecated in favor of general transaction.
 pub trait CreateInherent<LocalCall>: CreateTransactionBase<LocalCall> {
```
The naming is still just wrong and trying to fix this via docs is not making it better. As I already said in the original pr, the name of the trait is wrong. 

I would propose we introduce a new `CreateBare` with proper docs and tag this trait deprecated.

gui1117 replied:
I updated the PR: deprecate `CreateInherent`, introduce `CreateBare` and replace all usage in polkadot-sdk.

It is a breaking change so I am not sure I can backport anything to 2412, but I will backport it to 2503.

I acknowledge my mistake in transaction extension review, overall I underestimated the stability and the pace of new feature in polkadot-sdk in 2024.

As a side note I think offchain traits could be improved as well:
* we could remove `LocalCall` everywhere, people can bound `RuntimeCall: From<Call>` in their config
* `CreateTransactionBase` could simply bound `frame_system::Config` as supertrait instead of having a duplicated associated type `RuntimeCall` that bring name conflict. (We do bound `Config` as supertrait for `SigningTypes`).
But better not to break them.

georgepisaltu replied:
I agree with the offchain trait refactor, but it's probably best to do it after phase 2 of Extrinsic Horizon is done.

gui1117 commented on prdoc/pr_7597.prdoc:
```diff
@@ -0,0 +1,31 @@
+- audience: Runtime Dev
+  description: null
+crates:
+- name: frame-system
+  bump: major
+- name: polkadot-runtime-common
+  bump: major
+- name: polkadot-runtime-parachains
+  bump: major
+- name: rococo-runtime
+  bump: major
+- name: westend-runtime
+  bump: major
+- name: pallet-babe
+  bump: major
+- name: pallet-beefy
+  bump: major
+- name: pallet-election-provider-multi-block
+  bump: major
+- name: pallet-election-provider-multi-phase
+  bump: major
+- name: pallet-grandpa
+  bump: major
+- name: pallet-im-online
+  bump: major
+- name: pallet-mixnet
+  bump: major
+- name: pallet-offences-benchmarking
+  bump: major
```
```suggestion
- name: frame-system
  bump: major
- name: polkadot-runtime-common
  bump: patch
- name: polkadot-runtime-parachains
  bump: major
- name: rococo-runtime
  bump: major
- name: westend-runtime
  bump: major
- name: pallet-babe
  bump: major
- name: pallet-beefy
  bump: major
- name: pallet-election-provider-multi-block
  bump: major
- name: pallet-election-provider-multi-phase
  bump: major
- name: pallet-grandpa
  bump: major
- name: pallet-im-online
  bump: major
- name: pallet-mixnet
  bump: major
- name: pallet-offences-benchmarking
  bump: patch
```

gui1117 commented on substrate/frame/system/src/offchain.rs:
```diff
@@ -489,11 +489,22 @@ pub trait CreateSignedTransaction<LocalCall>:
 	/// Create an inherent.
 	fn create_inherent(call: Self::RuntimeCall) -> Self::Extrinsic;
 }
 
```
we can also automatically implement it:
pro: pallet using `CreateInherent` will still compile
cons: when people first bump `frame-system` version they will get a more confusing error:
```
warning: use of deprecated trait `polkadot_sdk::frame_system::offchain::CreateInherent`: Use `CreateBare` instead
    --> substrate/bin/node/runtime/src/lib.rs:1814:41
     |
1814 | impl<LocalCall> frame_system::offchain::CreateInherent<LocalCall> for Runtime
     |                                         ^^^^^^^^^^^^^^
     |
     = note: `#[warn(deprecated)]` on by default

error[E0277]: the trait bound `Runtime: CreateBare<polkadot_sdk::frame_system::Call<Runtime>>` is not satisfied
   --> substrate/bin/node/runtime/src/lib.rs:369:39
    |
369 | impl pallet_example_tasks::Config for Runtime {
    |                                       ^^^^^^^ the trait `CreateBare<polkadot_sdk::frame_system::Call<Runtime>>` is not implemented for `Runtime`
    |
note: required by a bound in `pallet_example_tasks::Config`
   --> /home/guillaume/remote-builds/polkadot-sdk/substrate/frame/examples/tasks/src/lib.rs:94:20
    |
94  |     pub trait Config: CreateBare<frame_system::Call<Self>> + frame_system::Config {
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Config`

```
IMO it is more confusing than helpful.

```suggestion

#[allow(deprecated)]
impl<T, C> CreateInherent<C> for T
where
	T: CreateBare<C>,
{
	fn create_inherent(call: Self::RuntimeCall) -> Self::Extrinsic {
		T::create_bare(call)
	}
}

```

bkchr replied:
Maybe we should do the inverse. See what I ahve written below and implement `CreateBare` for `CreateInherent`.

gui1117 commented:
/cmd fmt

bkchr commented on substrate/frame/system/src/offchain.rs:
```diff
@@ -571,7 +576,7 @@ pub trait SendSignedTransaction<
 
 /// Submit an unsigned transaction onchain with a signed payload
-pub trait SendUnsignedTransaction<T: SigningTypes + CreateInherent<LocalCall>, LocalCall> {
+pub trait SendUnsignedTransaction<T: SigningTypes + CreateBare<LocalCall>, LocalCall> {
```
If you change this, it basically means that people may have `CreateInherent` implemented, but it will not be used by anything.

bkchr replied:
Basically also resulting in compile errors.

gui1117 replied:
If we do this, runtime will have to implement `CreateInherent` with the intention of creating a bare extrinsic.

It is ok to me, and the least breaking change, but it keeps the faulty trait: `CreateInherent`, I will also add in the doc of `CreateInherent` the intention.

gui1117 replied:
I updated the code:
* `CreateInherent` has a new method `create_bare` with default implementation. User can override it to specify the behavior. (not ideal but not a breaking change.)
* `CreateBare` is introduced, automatically implemented for `CreateInherent` using `CreateInherent::create_bare`.
* `CreateBare` is used instead of `CreateInherent` for `SendUnsignedTransaction` and it implementation for `Signer`.
* pallets such as beefy are not using `CreateBare` in their trait bound to avoid any breaking change.
* User are expected to implement `CreateInherent` for their runtime in order to keep pallet's usage.
* User are expected to use `CreateBare` in their pallet, the idiomatic trait to create a bare extrinsic.

Possible alternative design: slightly more breaking, but provides a path forward:
* I considered deprecated `CreateInherent` and use `CreateBare` in polkadot-sdk pallets. That would be a breaking change: e.g. if a pallets bounds beefy config and use `T::create_inherent` method then their code then it won't compile. They have to use `create_bare` and import `CreateBase` trait in scope. (note: we could introduce `CreateBare::create_inherent` so that rustc give some hint to import the trait in scope.)

georgepisaltu replied:
I think it's ok as it is right now, fixes the name and helps the dev a bit with the default impl.

gui1117 commented:
/cmd fmt

bkchr commented on substrate/frame/system/src/offchain.rs:
```diff
@@ -489,9 +483,37 @@ pub trait CreateSignedTransaction<LocalCall>:
+///
+/// Implement this trait for the runtime but use `CreateBare` instead.
+/// This trait is defined to avoid a breaking change, and it automatically implements `CreateBare`.
 pub trait CreateInherent<LocalCall>: CreateTransactionBase<LocalCall> {
```
```suggestion
#[deprecated(note = "Please use `CreateBare` instead")]
pub trait CreateInherent<LocalCall>: CreateTransactionBase<LocalCall> {
```

gui1117 replied:
Then I will also change usage of `CreateInherent` to `CreateBare` in pokadot-sdk, in pallets and runtime

bkchr replied:
> EDIT: otherwise runtime implementing only `CreateBare` won't be able to use like `election-solution` pallet which uses `CreateInherent`.

I mean you can still implement `CreateInherent`, just with a `allow(deprecated)`.

georgepisaltu reviewed (approved):
If it only took a couple of months before full adoption of the new interface, this trait would eventually end up only being used to create inherents. We expected the transition period to be much shorter than what it seems it will be, which is a miss on my end definitely.

In order to prevent similar situations in the future, we should push to get out of this "limbo" stage where the feature is partially complete and unsigned extrinsics are still around but so are bare extrinsics and inherents. We should start by delivering phase 2 of [Extrinsic Horizon](https://github.com/paritytech/polkadot-sdk/issues/2415), implemented in https://github.com/paritytech/polkadot-sdk/pull/6324 https://github.com/paritytech/polkadot-sdk/pull/6325 https://github.com/paritytech/polkadot-sdk/pull/6326, to finally get rid of unsigned transactions and become consistent again regarding the terminology we use.

gui1117 commented:
It doesn't work either, because `CreateBare` takes a generic so we can't implement it directly when it is derived automatically for `CreateInherent`:

```
error[E0119]: conflicting implementations of trait `frame_system::offchain::CreateBare<_>` for type `mock::Test`
  --> substrate/frame/sassafras/src/mock.rs:59:1
   |
59 | / impl<C> frame_system::offchain::CreateBare<C> for Test
60 | | where
61 | |     RuntimeCall: From<C>,
   | |_________________________^
   |
   = note: conflicting implementation in crate `frame_system`:
           - impl<LocalCall, T> frame_system::offchain::CreateBare<LocalCall> for T
             where T: CreateInherent<LocalCall>;
   = note: downstream crates may implement trait `frame_system::offchain::CreateInherent<_>` for type `mock::Test`

For more information about this error, try `rustc --explain E0119`.
error: could not compile `pallet-sassafras` (lib test) due to 1 previous error
```

So instead I just renamed the trait `CreateInherent` to `CreateBare`, deprecated the method `create_inherent` and added the method `create_bare`, people updating their code will have a compile error saying `create_bare` is not implemented, I think it is fine because it is very easy to implement. Other than that there is no breaking change.

The deprecation of a reexport doesn't work in rust, but the doc show the rename so if people use an IDE they will see the trait is renamed.

![image](https://github.com/user-attachments/assets/165e3029-7791-4b11-bdf4-43dc372635f4)

gui1117 commented:
/cmd update-ui

gui1117 commented:
Considering approvals here for the PR when it was in a different state, I am not sure I should consider this PR approved.

But IMO it looks good.

gui1117 commented:
Having `CreateBare` as a trait automatically implemented for implementations `CreateInherent` is not doable because the trait is generic and it will fail with:

```
error[E0119]: conflicting implementations of trait `Bare<_>` for type `Runtime`
 --> src/lib.rs:3:1
  |
3 | impl<Call> major::Bare<Call> for Runtime {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: conflicting implementation in crate `major`:
          - impl<Call, T> Bare<Call> for T
            where T: Inherent<Call>;
  = note: downstream crates may implement trait `major::Inherent<_>` for type `Runtime`

For more information about this error, try `rustc --explain E0119`.
```

Having `CreateInherent` automatically implemented for `CreateBare` is doable but it is a breaking change, user will have to remove their implementation of `CreateInherent` and instead implement `CreateBare`.

In this PR, I choose a rename of the trait, I know renames are not very good but the documentation `CreateInherent` will show that it is renamed to `CreateBare` and "deprecated".

Another solution would simply to mark the trait as deprecated and create a new trait, and update all our usage for the new trait `CreateBare`. This will give a good deprecated warning message but also some compile error because `CreateBare` is not implemented.

If you have some opinion @bkchr

gui1117 commented:
@bkchr because you approve a different version that I had to change because it didn't work at the end.

This final version is:
A new trait `CreateBare` with 2 methods `create_bare` and deprecated `create_inherent` that default implementation is `create_bare`.
`CreateInherent` is rewrote as a type alias to `CreateBare` with a documentation saying it is deprecated.

User using `CreateInherent` will just have a warning `create_inherent` is deprecated in favor of `create_bare`.
Implementer of `CreateInherent` will have a compile error saying `create_bare` is not implemented.

I think it is good.
